### PR TITLE
Address TODO for UnicodePatternSyntaxChar check in Lexer

### LINKED ISSUE #25147: Address TODO's in the Lexer
**Description:**

$subject

### PR DESCRIPTION
**Pull request by lochana-chathura**
## Purpose
$subject.

Fixes #25147 

## Approach
N/A

## Samples
Earlier following identifier was allowed. Now those invalid identifiers are disallowed.  
```bal
int 😃invalidIdentifier = 10;
```

## Remarks
We seem to have no direct API to check Unicode property `Pattern_Syntax=True`.

As per the http://www.unicode.org/reports/tr31/tr31-22.html#Figure_Code_Point_Categories_for_Identifier_Parsing,

> Identifier characters, Pattern_Syntax characters, and Pattern_White_Space are disjoint: they will never overlap.

Therefore, can use [isUnicodeIdentifierPart()](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/Character.html#isUnicodeIdentifierPart(int)) to disallow Pattern_Syntax characters.

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [x] Added necessary documentation  
  - [x] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Review by KRVPerera — APPROVED** [review]
LGTM

File: ballerina-shell/modules/shell-core/src/test/resources/testcases/evaluator/basics.quoted.json [hunk 1]
```diff
@@ -45,11 +45,6 @@
     "code": "string 'üňĩćőđę_ňāɱȇ = \"John doe\"; 'üňĩćőđę_ňāɱȇ",
     "expr": "\"John doe\""
   },
-  {
-    "description": "Use unicode with escape in name.",
-    "code": "string 'unicode_\\u{2324} = \"Jane doe\"; 'unicode_\\u{2324}",
-    "expr": "\"Jane doe\""
-  },
```
Comment (lochana-chathura) [inline ballerina-shell/modules/shell-core/src/test/resources/testcases/evaluator/basics.quoted.json:52]: In the shell, we get 
`error: undefined symbol 'unicode_', 'unicode_⌤ // Variable is unable to be referenced after declaration` for this code sample. 

But, this sample compiles as expected in the compiler and  `'unicode_⌤` can be referenced after declaration.
Hence created #30021 for the shell issue.

**Review by rdulmina — APPROVED** [review]

File: ballerina-shell/modules/shell-core/src/test/resources/testcases/evaluator/basics.quoted.json (hunk 1 above)
Comment (lochana-chathura) [reply ballerina-shell/modules/shell-core/src/test/resources/testcases/evaluator/basics.quoted.json:52]: Reason is https://github.com/ballerina-platform/ballerina-lang/issues/30021#issuecomment-821957812

File: tests/jballerina-debugger-integration-test/src/test/resources/project-based-tests/variable-tests/main.bal [hunk 2]
```diff
@@ -144,9 +144,9 @@ public function main() {
     [int, string] tupleVar = [20, "foo"];
     map<string> mapVar = {line1: "No. 20", line2: "Palm Grove", city: "Colombo 03", country: "Sri Lanka"};
 
-    '\ \/\:\@\[\`\{\~\u{2324}_123_ƮέŞŢ_Student recordVar = {
+    '\ \/\:\@\[\`\{\~\u{03C0}_123_ƮέŞŢ_Student recordVar = {
         '1st_name: "John Doe",
-        'Ȧɢέ_\ \/\:\@\[\`\{\~⌤: 20,
+        'Ȧɢέ_\ \/\:\@\[\`\{\~π: 20,
```
Comment (lochana-chathura) [inline tests/jballerina-debugger-integration-test/src/test/resources/project-based-tests/variable-tests/main.bal:149]: Same reason as https://github.com/ballerina-platform/ballerina-lang/issues/30021#issuecomment-821957812

**Review by mohanvive — APPROVED** [review]

**Pull request merged**